File: compatibility-tests/sdk-test-java/src/test/java/com/floci/test/AcmTest.java
package com.floci.test;

import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.Assumptions;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.MethodOrderer;
import org.junit.jupiter.api.Order;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestMethodOrder;
import software.amazon.awssdk.core.SdkBytes;
import software.amazon.awssdk.services.acm.AcmClient;
import software.amazon.awssdk.services.acm.model.*;

import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

import static org.assertj.core.api.Assertions.*;

@DisplayName("ACM Certificate Manager")
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class AcmTest {

    private static AcmClient acm;
    private static final List<String> arnsToCleanup = new ArrayList<>();

    // Lifecycle test state
    private static String requestedCertArn;

    // Import/Export test state
    private static String importedCertArn;
    private static byte[] importedCertPem;
    private static byte[] importedKeyPem;
    private static String exportTestCertArn;

    // Tagging test state
    private static String tagTestCertArn;

    @BeforeAll
    static void setup() {
        acm = TestFixtures.acmClient();
    }

    @AfterAll
    static void cleanup() {
        if (acm != null) {
            for (String arn : arnsToCleanup) {
                try {
                    acm.deleteCertificate(b -> b.certificateArn(arn));
                } catch (Exception ignored) {}
            }
            acm.close();
        }
    }

    // ============================================
    // Helper: generate a self-signed certificate using openssl
    // ============================================
    private static byte[][] generateSelfSignedCert() throws Exception {
        Path keyFile = Files.createTempFile("test-key", ".pem");
        Path certFile = Files.createTempFile("test-cert", ".pem");
        try {
            ProcessBuilder pb = new ProcessBuilder("openssl", "req", "-x509", "-newkey", "rsa:2048",
                    "-keyout", keyFile.toString(), "-out", certFile.toString(),
                    "-days", "365", "-nodes", "-subj", "/CN=test.example.com");
            pb.redirectErrorStream(true);
            Process p = pb.start();
            p.getInputStream().readAllBytes(); // consume output
            int exitCode = p.waitFor();
            if (exitCode != 0) {
                throw new RuntimeException("openssl failed with exit code " + exitCode);
            }

            byte[] certPem = Files.readAllBytes(certFile);
            byte[] keyPem = Files.readAllBytes(keyFile);
            return new byte[][] { certPem, keyPem };
        } finally {
            Files.deleteIfExists(keyFile);
            Files.deleteIfExists(certFile);
        }
    }

    // ============================================
    // US1: Lifecycle tests
    // ============================================

    @Test
    @Order(1)
    @DisplayName("Request a certificate")
    void testRequestCertificate() {
        String domain = TestFixtures.uniqueName("java-test") + ".example.com";

        RequestCertificateResponse response = acm.requestCertificate(b -> b
                .domainName(domain));

        requestedCertArn = response.certificateArn();
        arnsToCleanup.add(requestedCertArn);

        assertThat(requestedCertArn).isNotNull();
        assertThat(requestedCertArn).matches("arn:aws:acm:.*:.*:certificate/.*");
    }

    @Test
    @Order(2)
    @DisplayName("Describe a certificate")
    void testDescribeCertificate() {
        Assumptions.assumeTrue(requestedCertArn != null, "RequestCertificate must succeed first");

        DescribeCertificateResponse response = acm.describeCertificate(b -> b
                .certificateArn(requestedCertArn));

        CertificateDetail detail = response.certificate();
        assertThat(detail.domainName()).contains("example.com");
        assertThat(detail.statusAsString()).isNotNull();
    }

    @Test
    @Order(3)
    @DisplayName("Get a certificate")
    void testGetCertificate() {
        Assumptions.assumeTrue(requestedCertArn != null, "RequestCertificate must succeed first");

        GetCertificateResponse response = acm.getCertificate(b -> b
                .certificateArn(requestedCertArn));

        assertThat(response.certificate()).isNotNull();
        assertThat(response.certificate()).contains("BEGIN CERTIFICATE");
    }

    @Test
    @Order(4)
    @DisplayName("List certificates contains the requested cert")
    void testListCertificates() {
        Assumptions.assumeTrue(requestedCertArn != null, "RequestCertificate must succeed first");

        ListCertificatesResponse response = acm.listCertificates();

        assertThat(response.certificateSummaryList())
                .anyMatch(s -> s.certificateArn().equals(requestedCertArn));
    }

    @Test
    @Order(5)
    @DisplayName("Delete a certificate")
    void testDeleteCertificate() {
        Assumptions.assumeTrue(requestedCertArn != null, "RequestCertificate must succeed first");

        acm.deleteCertificate(b -> b.certificateArn(requestedCertArn));
        arnsToCleanup.remove(requestedCertArn);

        assertThatThrownBy(() -> acm.describeCertificate(b -> b
                .certificateArn(requestedCertArn)))
                .isInstanceOf(AcmException.class);
    }

    // ============================================
    // US2: Import/Export tests
    // ============================================

    @Test
    @Order(10)
    @DisplayName("Import a self-signed certificate")
    void testImportCertificate() throws Exception {
        byte[][] certAndKey = generateSelfSignedCert();
        importedCertPem = certAndKey[0];
        importedKeyPem = certAndKey[1];

        ImportCertificateResponse response = acm.importCertificate(b -> b
                .certificate(SdkBytes.fromByteArray(importedCertPem))
                .privateKey(SdkBytes.fromByteArray(importedKeyPem)));

        importedCertArn = response.certificateArn();
        arnsToCleanup.add(importedCertArn);

        assertThat(importedCertArn).isNotNull();
        assertThat(importedCertArn).matches("arn:aws:acm:.*:.*:certificate/.*");
    }

    @Test
    @Order(11)
    @DisplayName("Get an imported certificate")
    void testGetImportedCertificate() {
        Assumptions.assumeTrue(importedCertArn != null, "ImportCertificate must succeed first");

        GetCertificateResponse response = acm.getCertificate(b -> b
                .certificateArn(importedCertArn));

        assertThat(response.certificate()).isNotNull();
        assertThat(response.certificate()).contains("BEGIN CERTIFICATE");
    }

    @Test
    @Order(12)
    @DisplayName("Export an imported certificate with passphrase")
    void testExportCertificate() {
        Assumptions.assumeTrue(importedCertArn != null, "ImportCertificate must succeed first");

        SdkBytes passphrase = SdkBytes.fromUtf8String("test-passphrase-123");

        ExportCertificateResponse response = acm.exportCertificate(b -> b
                .certificateArn(importedCertArn)
                .passphrase(passphrase));

        assertThat(response.certificate()).isNotNull();
        assertThat(response.certificate()).contains("BEGIN CERTIFICATE");
        assertThat(response.privateKey()).isNotNull();
        assertThat(response.privateKey()).contains("-----BEGIN");
    }

    @Test
    @Order(13)
    @DisplayName("Export a requested (non-imported) certificate fails")
    void testExportRequestedCertificateFails() {
        // Request a new cert (not imported, so no private key to export)
        String domain = TestFixtures.uniqueName("export-fail") + ".example.com";
        RequestCertificateResponse reqResp = acm.requestCertificate(b -> b.domainName(domain));
        exportTestCertArn = reqResp.certificateArn();
        arnsToCleanup.add(exportTestCertArn);

        SdkBytes passphrase = SdkBytes.fromUtf8String("test-passphrase");

        assertThatThrownBy(() -> acm.exportCertificate(b -> b
                .certificateArn(exportTestCertArn)
                .passphrase(passphrase)))
                .isInstanceOf(AcmException.class);
    }

    // ============================================
    // US3: Tagging tests
    // ============================================

    @Test
    @Order(20)
    @DisplayName("Add and list tags on a certificate")
    void testAddTags() {
        String domain = TestFixtures.uniqueName("tag-test") + ".example.com";
        RequestCertificateResponse reqResp = acm.requestCertificate(b -> b.domainName(domain));
        tagTestCertArn = reqResp.certificateArn();
        arnsToCleanup.add(tagTestCertArn);

        acm.addTagsToCertificate(b -> b
                .certificateArn(tagTestCertArn)
                .tags(
                        software.amazon.awssdk.services.acm.model.Tag.builder().key("Environment").value("test").build(),
                        software.amazon.awssdk.services.acm.model.Tag.builder().key("Project").value("floci").build()
                ));

        ListTagsForCertificateResponse tagsResp = acm.listTagsForCertificate(b -> b
                .certificateArn(tagTestCertArn));

        assertThat(tagsResp.tags()).hasSize(2);
        assertThat(tagsResp.tags())
                .anyMatch(t -> t.key().equals("Environment") && t.value().equals("test"));
        assertThat(tagsResp.tags())
                .anyMatch(t -> t.key().equals("Project") && t.value().equals("floci"));
    }

    @Test
    @Order(21)
    @DisplayName("Remove a tag from a certificate")
    void testRemoveTags() {
        Assumptions.assumeTrue(tagTestCertArn != null, "AddTags must succeed first");

        acm.removeTagsFromCertificate(b -> b
                .certificateArn(tagTestCertArn)
                .tags(software.amazon.awssdk.services.acm.model.Tag.builder().key("Project").value("floci").build()));

        ListTagsForCertificateResponse tagsResp = acm.listTagsForCertificate(b -> b
                .certificateArn(tagTestCertArn));

        assertThat(tagsResp.tags()).hasSize(1);
        assertThat(tagsResp.tags())
                .anyMatch(t -> t.key().equals("Environment") && t.value().equals("test"));
        assertThat(tagsResp.tags())
                .noneMatch(t -> t.key().equals("Project"));
    }

    // ============================================
    // US4: Account configuration tests
    // ============================================

    @Test
    @Order(30)
    @DisplayName("Put and get account configuration")
    void testPutAndGetAccountConfiguration() {
        acm.putAccountConfiguration(b -> b
                .expiryEvents(e -> e.daysBeforeExpiry(45))
                .idempotencyToken(TestFixtures.uniqueName()));

        var response = acm.getAccountConfiguration(b -> b.build());

        assertThat(response.expiryEvents()).isNotNull();
        assertThat(response.expiryEvents().daysBeforeExpiry()).isEqualTo(45);
    }

    // ============================================
    // US5: Error handling tests
    // ============================================

    @Test
    @Order(40)
    @DisplayName("Describe a non-existent certificate throws exception")
    void testDescribeNonExistentCertificate() {
        String fakeArn = "arn:aws:acm:us-east-1:000000000000:certificate/00000000-0000-0000-0000-000000000000";

        assertThatThrownBy(() -> acm.describeCertificate(b -> b
                .certificateArn(fakeArn)))
                .isInstanceOf(AcmException.class);
    }

    @Test
    @Order(41)
    @DisplayName("Request certificate with SANs")
    void testRequestWithSANs() {
        String domain = TestFixtures.uniqueName("san-test") + ".example.com";
        String san1 = "www." + domain;
        String san2 = "api." + domain;

        RequestCertificateResponse response = acm.requestCertificate(b -> b
                .domainName(domain)
                .subjectAlternativeNames(san1, san2));

        String arn = response.certificateArn();
        arnsToCleanup.add(arn);
        assertThat(arn).isNotNull();

        DescribeCertificateResponse descResp = acm.describeCertificate(b -> b
                .certificateArn(arn));

        assertThat(descResp.certificate().subjectAlternativeNames())
                .contains(san1, san2);
    }

    @Test
    @Order(42)
    @DisplayName("Import invalid PEM throws exception")
    void testImportInvalidPEM() {
        byte[] garbage = "this is not a valid certificate".getBytes();

        assertThatThrownBy(() -> acm.importCertificate(b -> b
                .certificate(SdkBytes.fromByteArray(garbage))
                .privateKey(SdkBytes.fromByteArray(garbage))))
                .isInstanceOf(AcmException.class);
    }
}

File: compatibility-tests/sdk-test-node/tests/acm.test.ts
/**
 * ACM integration tests.
 */

import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import {
  ACMClient,
  RequestCertificateCommand,
  DescribeCertificateCommand,
  GetCertificateCommand,
  ListCertificatesCommand,
  DeleteCertificateCommand,
  ImportCertificateCommand,
  ExportCertificateCommand,
  AddTagsToCertificateCommand,
  ListTagsForCertificateCommand,
  RemoveTagsFromCertificateCommand,
  PutAccountConfigurationCommand,
  GetAccountConfigurationCommand,
} from '@aws-sdk/client-acm';
import selfsigned from 'selfsigned';
import { makeClient, uniqueName, ACCOUNT, REGION } from './setup';

async function generateSelfSignedCert(): Promise<{ cert: Buffer; key: Buffer }> {
  const attrs = [{ name: 'commonName', value: 'test.example.com' }];
  const pems = await selfsigned.generate(attrs);
  return { cert: Buffer.from(pems.cert), key: Buffer.from(pems.private) };
}

describe('ACM Certificate Lifecycle', () => {
  let acm: ACMClient;
  let certificateArn: string;

  beforeAll(() => {
    acm = makeClient(ACMClient);
  });

  afterAll(async () => {
    try {
      if (certificateArn) {
        await acm.send(new DeleteCertificateCommand({ CertificateArn: certificateArn }));
      }
    } catch {
      // ignore
    }
  });

  it('should request a certificate', async () => {
    const domain = `${uniqueName('cert')}.example.com`;
    const response = await acm.send(
      new RequestCertificateCommand({ DomainName: domain })
    );
    certificateArn = response.CertificateArn!;
    expect(certificateArn).toMatch(
      new RegExp(`^arn:aws:acm:${REGION}:${ACCOUNT}:certificate/.+`)
    );
  });

  it('should describe a certificate', async () => {
    const response = await acm.send(
      new DescribeCertificateCommand({ CertificateArn: certificateArn })
    );
    expect(response.Certificate?.DomainName).toBeTruthy();
    expect(response.Certificate?.Status).toBe('ISSUED');
  });

  it('should get certificate', async () => {
    const response = await acm.send(
      new GetCertificateCommand({ CertificateArn: certificateArn })
    );
    expect(response.Certificate).toBeTruthy();
  });

  it('should list certificates', async () => {
    const response = await acm.send(new ListCertificatesCommand({}));
    expect(
      response.CertificateSummaryList?.some((c) => c.CertificateArn === certificateArn)
    ).toBe(true);
  });

  it('should delete a certificate', async () => {
    await acm.send(new DeleteCertificateCommand({ CertificateArn: certificateArn }));

    await expect(
      acm.send(new DescribeCertificateCommand({ CertificateArn: certificateArn }))
    ).rejects.toThrow();

    // Already deleted; prevent afterAll from trying again
    certificateArn = '';
  });
});

describe('ACM Import/Export', () => {
  let acm: ACMClient;
  const arnsToClean: string[] = [];

  beforeAll(() => {
    acm = makeClient(ACMClient);
  });

  afterAll(async () => {
    for (const arn of arnsToClean) {
      try {
        await acm.send(new DeleteCertificateCommand({ CertificateArn: arn }));
      } catch {
        // ignore
      }
    }
  });

  it('should import a certificate', async () => {
    const { cert, key } = await generateSelfSignedCert();
    const response = await acm.send(
      new ImportCertificateCommand({
        Certificate: cert,
        PrivateKey: key,
      })
    );
    const arn = response.CertificateArn!;
    arnsToClean.push(arn);
    expect(arn).toMatch(
      new RegExp(`^arn:aws:acm:${REGION}:${ACCOUNT}:certificate/.+`)
    );
  });

  it('should get imported certificate', async () => {
    const { cert, key } = await generateSelfSignedCert();
    const importResp = await acm.send(
      new ImportCertificateCommand({
        Certificate: cert,
        PrivateKey: key,
      })
    );
    const arn = importResp.CertificateArn!;
    arnsToClean.push(arn);

    const getResp = await acm.send(
      new GetCertificateCommand({ CertificateArn: arn })
    );
    expect(getResp.Certificate).toBeTruthy();
  });

  it('should export imported certificate', async () => {
    const { cert, key } = await generateSelfSignedCert();
    const importResp = await acm.send(
      new ImportCertificateCommand({
        Certificate: cert,
        PrivateKey: key,
      })
    );
    const arn = importResp.CertificateArn!;
    arnsToClean.push(arn);

    const passphrase = Buffer.from('test-passphrase');
    const exportResp = await acm.send(
      new ExportCertificateCommand({
        CertificateArn: arn,
        Passphrase: passphrase,
      })
    );
    expect(exportResp.Certificate).toBeTruthy();
    expect(exportResp.PrivateKey).toBeTruthy();
  });

  it('should fail to export requested certificate', async () => {
    const domain = `${uniqueName('exp')}.example.com`;
    const reqResp = await acm.send(
      new RequestCertificateCommand({ DomainName: domain })
    );
    const arn = reqResp.CertificateArn!;
    arnsToClean.push(arn);

    await expect(
      acm.send(
        new ExportCertificateCommand({
          CertificateArn: arn,
          Passphrase: Buffer.from('test-passphrase'),
        })
      )
    ).rejects.toThrow();
  });
});

describe('ACM Tagging', () => {
  let acm: ACMClient;
  let certificateArn: string;

  beforeAll(async () => {
    acm = makeClient(ACMClient);
    const domain = `${uniqueName('tag')}.example.com`;
    const response = await acm.send(
      new RequestCertificateCommand({ DomainName: domain })
    );
    certificateArn = response.CertificateArn!;
  });

  afterAll(async () => {
    try {
      if (certificateArn) {
        await acm.send(new DeleteCertificateCommand({ CertificateArn: certificateArn }));
      }
    } catch {
      // ignore
    }
  });

  it('should add and list tags', async () => {
    await acm.send(
      new AddTagsToCertificateCommand({
        CertificateArn: certificateArn,
        Tags: [
          { Key: 'Environment', Value: 'test' },
          { Key: 'Project', Value: 'floci' },
        ],
      })
    );

    const response = await acm.send(
      new ListTagsForCertificateCommand({ CertificateArn: certificateArn })
    );
    const tags = response.Tags ?? [];
    expect(tags.some((t) => t.Key === 'Environment' && t.Value === 'test')).toBe(true);
    expect(tags.some((t) => t.Key === 'Project' && t.Value === 'floci')).toBe(true);
  });

  it('should remove tags', async () => {
    // Ensure tags exist
    await acm.send(
      new AddTagsToCertificateCommand({
        CertificateArn: certificateArn,
        Tags: [{ Key: 'ToRemove', Value: 'yes' }],
      })
    );

    await acm.send(
      new RemoveTagsFromCertificateCommand({
        CertificateArn: certificateArn,
        Tags: [{ Key: 'ToRemove', Value: 'yes' }],
      })
    );

    const response = await acm.send(
      new ListTagsForCertificateCommand({ CertificateArn: certificateArn })
    );
    const tags = response.Tags ?? [];
    expect(tags.some((t) => t.Key === 'ToRemove')).toBe(false);
  });
});

describe('ACM Account Configuration', () => {
  let acm: ACMClient;

  beforeAll(() => {
    acm = makeClient(ACMClient);
  });

  it('should put and get account configuration', async () => {
    await acm.send(
      new PutAccountConfigurationCommand({
        ExpiryEvents: { DaysBeforeExpiry: 45 },
        IdempotencyToken: uniqueName('idem'),
      })
    );

    const response = await acm.send(new GetAccountConfigurationCommand({}));
    expect(response.ExpiryEvents?.DaysBeforeExpiry).toBe(45);
  });
});

describe('ACM Error Handling', () => {
  let acm: ACMClient;
  const arnsToClean: string[] = [];

  beforeAll(() => {
    acm = makeClient(ACMClient);
  });

  afterAll(async () => {
    for (const arn of arnsToClean) {
      try {
        await acm.send(new DeleteCertificateCommand({ CertificateArn: arn }));
      } catch {
        // ignore
      }
    }
  });

  it('should throw for non-existent certificate', async () => {
    const fakeArn = `arn:aws:acm:${REGION}:${ACCOUNT}:certificate/00000000-0000-0000-0000-000000000000`;
    await expect(
      acm.send(new DescribeCertificateCommand({ CertificateArn: fakeArn }))
    ).rejects.toThrow();
  });

  it('should handle request with SANs', async () => {
    const domain = `${uniqueName('san')}.example.com`;
    const san = `alt.${domain}`;
    const response = await acm.send(
      new RequestCertificateCommand({
        DomainName: domain,
        SubjectAlternativeNames: [domain, san],
      })
    );
    const arn = response.CertificateArn!;
    arnsToClean.push(arn);
    expect(arn).toBeTruthy();

    const descResp = await acm.send(
      new DescribeCertificateCommand({ CertificateArn: arn })
    );
    const sans = descResp.Certificate?.SubjectAlternativeNames ?? [];
    expect(sans).toContain(san);
  });

  it('should fail to import invalid PEM', async () => {
    await expect(
      acm.send(
        new ImportCertificateCommand({
          Certificate: Buffer.from('not-a-valid-pem'),
          PrivateKey: Buffer.from('also-not-valid'),
        })
      )
    ).rejects.toThrow();
  });
});
